Generalize VAST plugin for all ZeroMQ apps (#46)
tenzir/threatbus: plugins/apps/threatbus_zmq_app/README.md

Threat Bus App Plugin for ZMQ Protocol
======================================

<h4 align="center">

[![PyPI Status][pypi-badge]][pypi-url]
[![Build Status][ci-badge]][ci-url]
[![License][license-badge]][license-url]

</h4>

A Threat Bus plugin that enables communication with any application that can
communicate via [ZeroMQ].

## Installation

```sh
pip install threatbus-vast
```

## Configuration

The plugin uses ZeroMQ to communicate with applications, like the
[VAST bridge](https://github.com/tenzir/threatbus/tree/master/apps/vast). The
plugin serves three ZeroMQ endpoints to connect with. One endpoint for managing

> subscriptions (and thus snapshot requests). The other two endpoints are plainly
> for pub-sub operations.

Reviewer: ```suggestion
subscriptions (and thus snapshot requests). The other two endpoints exist
for pub-sub operations.
```

The reader wonders: one for publish and one for subscribe?

Author: Yes, that is how it's done with ZeroMQ.

```yaml
...
plugins:

>   vast:

Reviewer: Hasn't this changed?

Author: Yes! fixed it.

    host: "127.0.0.1"
        manage: 13370
        pub: 13371
        sub: 13372
...
```

Initially, apps that want to connect with this plugin only need to know the
`manage` endpoint. The plugin and the app negotiate all other internals for
pub-sub message exchange at runtime. See the protocol definition below for
details.

## Management Protocol

Subscriptions and unsubscriptions are referred to as `management`. All
management messages are JSON formatted and exchanged via the `manage` ZMQ

> endpoint, that the plugin exposes.

Reviewer: ```suggestion
endpoint that the plugin exposes.
```

The manage endpoint uses the
[ZeroMQ Request/Reply](https://learning-0mq-with-pyzmq.readthedocs.io/en/latest/pyzmq/patterns/client_server.html)
pattern for message exchange. That means, all messages get an immediate response
from the server. With each JSON reply, the zmq-app Threat Bus plugin sends a
`status` field that indicates the success of the requested operation.

### Subscribe at the Plugin

To subscribe to Threat Bus via the zmq-app plugin, an app needs to send a JSON
struct as follows to the `manage` endpoint of the plugin:

```
{
  "action": "subscribe",
  "topic": <TOPIC>,       # either 'threatbus/sighting' or 'threatbus/intel'
  "snapshot": <SNAPSHOT>  # number of days for a snapshot, 0 for no snapshot
}
```

> In response, the app will either receive a `success` or `unsuccess` response.

Reviewer: ```suggestion
In response, the app will either receive a `success` or `error` response.
```

Any reason to not use this less obscure terminology?

Author: Changed it to `error` responses everywhere in the code base.

- Unsuccess response:
  ```
  {
    "status": "unsuccess"
  }
  ```
- Success response:
  ```
  {
    "topic": <P2P_TOPIC>,
    "pub_endpoint": 13371,
    "sub_endpoint": 13372,
    "status": "success",
  }
  ```

>   The `pub_endpoint` and `sub_endpoint` are the port numbers that are configured

Reviewer: An endpoint usually has the form `host:port`. Why not call it `pub_port`?

Author: Good catch, these are indeed endpoints and not port numbers, so the `pub_endpoint` and `sub_endpoint` are exactly in the form `host:port`. I'll update the docs.

  in the Threat Bus plugin config. The app can access these endpoints following
  the [ZeroMQ Pub/Sub](https://learning-0mq-with-pyzmq.readthedocs.io/en/latest/pyzmq/patterns/pubsub.html)
  pattern. The plugin will publish messages on the `pub_endpoint` and accept
  messages on the `sub_endpoint`.

  The `topic` field of the response contains a unique topic for the client. That
  topic _must_ be used to receive messages. The unique topic is a 32 characters
  wide random string and guarantees that other subscribers won't accidentally
  see traffic that should only be visible to the new client.
  
  For more details see below at `Pub/Sub via ZeroMQ`.

### Unsubscribe from the Plugin

> To unsubscribe a connected app, it has to send a JSON struct to the `manage`

Reviewer: Please plugin the subject for "it".

Author: Changed the sentence.

endpoint of the plugin, as follows:

```
{
  "action": "unsubscribe",
  "topic": <P2P_TOPIC>       # the 32-characters random topic that the got during subscription handshake
}
```

In response, the app will either receive a `success` or `unsuccess` response.

- Unsuccess response:
  ```
  {
    "status": "unsuccess"
  }
  ```
- Success response:
  ```
  {
    "status": "success"
  }
  ```

### Pub/Sub via ZeroMQ

Once an app has subscribed to Threat Bus via using the `manage` endpoint of the

> zmq-app plugin, it got a unique, random `p2p_topic` (see above). Threat Bus

Reviewer: ```suggestion
zmq-app plugin, it has a unique, random `p2p_topic` (see above). Threat Bus
```

For technical documentation, let's be precise as possible and avoid jargon if we can.

> (the zmq-app) uses this topic to publish all kinds of messages to the plugin, be

Reviewer: "all kinds" sounds very vague to me. Can you be more specific?

Author: I wrote out the message types and clarified the sentence.

it `Intel`, `Sightings` or `SnapshotRequests`.

ZeroMQ uses [prefix matching](https://zeromq.org/socket-api/#topics) for pub/sub
connections. The zmq-app plugin leverages this feature to indicate the type of
each sent message to the subscriber. Hence, app can simply match the topic
suffix to determine the message type.

For example, all `Intel` items will always be sent on the topic
`p2p_topic + "intel"`, all `SnapshotRequest` items on the topic
`p2p_topic + "snapshotrequest"` and so forth.

### Snapshots

Threat Bus' [snapshot](https://docs.tenzir.com/threatbus/features/snapshotting)
feature allows apps to request a snapshot during subscription. But snapshots

> also should be provided by apps when Threat Bus requests it. This is done via

Reviewer: "should" or will be provided? If this is a well-defined operation, it should be expressed with more rigor.

Author: Apps don't have to implement that feature, if they don't want to. E.g., the vast bridge does not implement it, but the MISP plugin does.

Author: I updated the text for clarity.

The plugin uses ZeroMQ to communicate with applications, like the
[VAST bridge](https://github.com/tenzir/threatbus/tree/master/apps/vast). The
plugin serves three ZeroMQ endpoints to connect with. One endpoint for managing
subscriptions (and thus snapshot requests). The other two endpoints exist
for pub-sub operations.

```yaml
...
plugins:
  zmq-app:
    host: "127.0.0.1"
        manage: 13370

>         pub: 13371

Reviewer: The hierarchy doesn't match the example config.
